base/iobuffer.jl

```diff
@@ -489,6 +489,62 @@ function take!(io::IOBuffer)
     return data
 end
 
+"Internal method. Copies the data from the IOBuffer to a string, leaving the
+IOBuffer in an unusable (erroneous) state.
+This assumes the IOBuffer is writable and seekable"
+function unsafe_takestring!(io::IOBuffer)
+    off = io.offset
+    nbytes = io.size - off
+    iszero(nbytes) && return ""
+    mem = StringMemory(nbytes)
+    unsafe_copyto!(mem, 1, io.data, off+1, nbytes)
+    return unsafe_takestring(mem)
+end
+
+"""
+    takestring!(io::IOBuffer) -> String
+
+Return the content of `io` as a `String`, resetting the buffer to its initial
+state.
+This is preferred over calling `String(take!(io))` to create a string from
+an `IOBuffer`.
+
+# Examples
+```jldoctest
+julia> io = IOBuffer();
+
+julia> write(io, [0x61, 0x62, 0x63]);
+
+julia> s = takestring!(io)
+"abc"
+
+julia> isempty(take!(io)) # io is now empty
+true
+```
+"""
+function takestring!(io::IOBuffer)
```

Comment:
This is missing a method for `GenericIOBuffer`.

Also, it is doing an extra copy that can sometimes be avoided.
Maybe the `IOBuffer` changes can be moved to a follow-up PR to make this PR easier to review?

Reply:
Yes, I like having this function but it copies unconditionally, which defeats the purpose. If you're going to copy the data there's no need for any unsafe or `!`.